File: lasso.jl
""" Regularization for Lasso Regression """
struct L1Regularization
    alpha :: Float64
end
    
function(self :: L1Regularization)( w :: Vector{Float64})
    return self.alpha * norm(w)
end

function grad(self :: L1Regularization, w :: Vector{Float64})
    return self.alpha .* sign.(w)
end

"""
    Linear regression model with a regularization factor which does both variable selection 
    and regularization. Model that tries to balance the fit of the model with respect to the training 
    data and the complexity of the model. A large regularization factor with decreases the variance of 
    the model and do para.

    # Parameters:
    
    - `degree`: int
        The degree of the polynomial that the independent variable X will be transformed to.
    - `reg_factor`: float
        The factor that will determine the amount of regularization and feature
        shrinkage. 
    - `n_iterations`: int
        The number of training iterations the algorithm will tune the weights for.
    - `learning_rate`: float
        The step length that will be used when updating the weights.
"""
struct LassoRegression

    degree         :: Int64
    regularization :: L1Regularization
    n_iterations   :: Int64
    learning_rate  :: Float64
    w              :: Vector{Float64}

    function LassoRegression( ; degree, reg_factor, n_features )

        n_iterations  = 3000
        learning_rate = 0.01
        w             = zeros(Float64, n_features) 
 
        new( degree, L1Regularization(reg_factor), n_iterations, 
             learning_rate, w)

    end

    function LassoRegression( ; degree, reg_factor, n_features, 
                              learning_rate, n_iterations)

        w  = zeros(Float64, n_features) 
        new( degree, L1Regularization(reg_factor), n_iterations, 
             learning_rate, w)

    end

end

function fit(self, X, y)
    X = normalize(polynomial_features(X, self.degree))
    # Insert constant ones for bias weights
    X = hcat( ones(eltype(X), size(X)[1]), X)
    self.training_errors = Float64[]
    n_features = size(X)[2]

    # Initialize weights randomly [-1/N, 1/N]
    limit = 1 / sqrt(n_features)
    self.w .= -limit + 2 .* limit .* rand(n_features)

    # Do gradient descent for n_iterations
    for i in 1:self.n_iterations
        y_pred = X * self.w
        # Calculate l2 loss
        mse = mean(0.5 .* (y .- y_pred).^2 + regularization(self.w))
        psuh!(self.training_errors, mse)
        # Gradient of l2 loss w.r.t w
        grad_w  = - (y .- y_pred) * X .+ grad(regularization, self.w)
        # Update the weights
        self.w .-= self.learning_rate .* grad_w
    end

end

function predict(self, X)
    X = normalize(polynomial_features(X, self.degree))
    # Insert constant ones for bias weights
    X = hcat( ones(eltype(X), size(X)[1]), X)
    y_pred = X * self.w
    return y_pred
end

File: lasso.py
#
# https://github.com/eriklindernoren/ML-From-Scratch
#

import numpy as np
import math
from utils import normalize, polynomial_features

class l1_regularization():
    """ Regularization for Lasso Regression """
    def __init__(self, alpha):
        self.alpha = alpha
    
    def __call__(self, w):
        return 

    def grad(self, w):
        return 

class LassoRegression():
    """
    Linear regression model with a regularization factor which does both variable selection 
    and regularization. Model that tries to balance the fit of the model with respect to the training 
    data and the complexity of the model. A large regularization factor with decreases the variance of 
    the model and do para.
    Parameters:
    -----------
    degree: int
        The degree of the polynomial that the independent variable X will be transformed to.
    reg_factor: float
        The factor that will determine the amount of regularization and feature
        shrinkage. 
    n_iterations: float
        The number of training iterations the algorithm will tune the weights for.
    learning_rate: float
        The step length that will be used when updating the weights.
    """

    def __init__(self, degree, reg_factor, n_iterations=3000, learning_rate=0.01):
        self.degree = degree
        self.alpha = reg_factor
        self.n_iterations = n_iterations
        self.learning_rate = learning_rate

    def fit(self, X, y):
        X = normalize(polynomial_features(X, degree=self.degree))
        # Insert constant ones for bias weights
        X = np.insert(X, 0, 1, axis=1)
        self.training_errors = []
        n_features=X.shape[1]

        # Initialize weights randomly [-1/N, 1/N]
        limit = 1 / math.sqrt(n_features)
        self.w = np.random.uniform(-limit, limit, (n_features, ))

        # Do gradient descent for n_iterations
        for i in range(self.n_iterations):
            y_pred = X.dot(self.w)
            # Calculate l2 loss
            mse = np.mean(0.5 * (y - y_pred)**2 + self.alpha * np.linalg.norm(self.w))
            self.training_errors.append(mse)
            # Gradient of l2 loss w.r.t w
            grad_w = -(y - y_pred).dot(X) + self.alpha * np.sign(self.w)
            # Update the weights
            self.w -= self.learning_rate * grad_w

    def predict(self, X):
        X = normalize(polynomial_features(X, degree=self.degree))
        # Insert constant ones for bias weights
        X = np.insert(X, 0, 1, axis=1)
        y_pred = X.dot(self.w)
        return y_pred


